File: koodit/computeSpectSquareShifts.cpp
template<typename T>
inline std::vector<std::vector<T>> computeSpectSquareShifts(paramStruct<T>& param, Det<T> detectors, int64_t ix, int64_t iy) {
	// Panel outer normal vector
	T panelNvecX = detectors.xs - detectors.xd; // Unnormalized detector panel normal vector X component
	T panelNvecY = detectors.ys - detectors.yd; // Unnormalized detector panel normal vector Y component
	T panelNvecXn = panelNvecX / std::sqrt(std::pow(panelNvecX, 2) + std::pow(panelNvecY, 2)); // Normalized ...
	T panelNvecYn = panelNvecY / std::sqrt(std::pow(panelNvecX, 2) + std::pow(panelNvecY, 2)); // Normalized ...

	// Panel rotation angle
	T panelAngle = atan2d(panelNvecY, panelNvecX);
	
	// Get detector pixel center in local coordinates from ix, iy and size
	T pixelCenterX = ix * param.dPitchXY - param.Nx * param.dPitchXY / 2. + param.dPitchXY / 2.;
	T pixelCenterY = iy * param.dPitchXY - param.Ny * param.dPitchXY / 2. + param.dPitchXY / 2.;

	// Pixel boundary in local coordinates
	T xdMin = pixelCenterX - param.dPitchXY / (T)2.;
	T xdMax = pixelCenterX + param.dPitchXY / (T)2.;
	T ydMin = pixelCenterY - param.dPitchXY / (T)2.;
	T ydMax = pixelCenterY + param.dPitchXY / (T)2.;
	T xsMin = xdMin - param.colD;
	T xsMax = xdMax + param.colD;
	T ysMin = ydMin - param.colD;
	T ysMax = ydMax + param.colD;

	T nRay = std::pow(std::ceil(std::sqrt(param.nRaySPECT)), 2);
	std::vector<std::vector<T>> detectorShifts(nRay, std::vector<T>(2, 0));; // 2D position shifts in detector end (for one hexagon)
	std::vector<std::vector<T>> sourceShifts(nRay, std::vector<T>(2, 0));; // 2D position shifts in source end (for one hexagon)

	for (uint32_t x = 0; x < std::sqrt(nRay); x++) {
		for (uint32_t y = 0; y < std::sqrt(nRay); y++) {
			T tmpXd = xdMin + x / (std::sqrt(nRay) - 1) * (xdMax - xdMin);
			T tmpYd = ydMin + y / (std::sqrt(nRay) - 1) * (ydMax - ydMin);
			T tmpXs = xsMin + x / (std::sqrt(nRay) - 1) * (xsMax - xsMin);
			T tmpYs = ysMin + y / (std::sqrt(nRay) - 1) * (ysMax - ysMin);

			detectorShifts[x * std::sqrt(nRay) + y][0] = tmpXd;
			detectorShifts[x * std::sqrt(nRay) + y][1] = tmpYd;
			sourceShifts[x * std::sqrt(nRay) + y][0] = tmpXs;
			sourceShifts[x * std::sqrt(nRay) + y][1] = tmpYs;
		}	
	}

	std::vector<std::vector<T>> sqShifts(nRay, std::vector<T>(6, 0));
	for (uint32_t currentShift = 0; currentShift < nRay; currentShift++) {
		T xx_d = detectorShifts[currentShift][0];
		T yy_d = detectorShifts[currentShift][1];
		T xx_s = sourceShifts[currentShift][0];
		T yy_s = sourceShifts[currentShift][1];

		sqShifts[currentShift][0] = -sind(panelAngle) * (yy_s-pixelCenterY) - panelNvecX - panelNvecXn * param.colL;
		sqShifts[currentShift][1] = cosd(panelAngle) * (yy_s-pixelCenterY) - panelNvecY - panelNvecYn * param.colL;
		sqShifts[currentShift][2] = xx_s - pixelCenterX;
		sqShifts[currentShift][3] = -sind(panelAngle) * (yy_d-pixelCenterY);
		sqShifts[currentShift][4] = cosd(panelAngle) * (yy_d-pixelCenterY);
		sqShifts[currentShift][5] = xx_d - pixelCenterX;
	}

	return sqShifts;
}

File: koodit/computeSpectSquareShifts.m
function sqShifts = computeSpectSquareShifts( ...
    param, ...
    detectors, ...
    ix, ...
    iy, ...
    twoDflag ...
)
    panelNvecX = detectors.xs - detectors.xd; % Unnormalized detector panel normal vector X component
    panelNvecY = detectors.ys - detectors.yd; % Unnormalized detector panel normal vector Y component
    panelNvecXn = panelNvecX / sqrt(panelNvecX.^2 + panelNvecY.^2); % Normalized detector panel normal vector X component
    panelNvecYn = panelNvecY / sqrt(panelNvecX.^2 + panelNvecY.^2); % Normalized detector panel normal vector Y component
    
    % Panel rotation angle
	panelAngle = atan2d(panelNvecY, panelNvecX);

    % Get detector pixel center in local coordinates from ix, iy and size
	pixelCenterX = ix * param.dPitchXY - param.Nx * param.dPitchXY / 2. + param.dPitchXY / 2.;
	pixelCenterY = iy * param.dPitchXY - param.Ny * param.dPitchXY / 2. + param.dPitchXY / 2.;

    % Pixel boundary in local coordinates
    xdMin = pixelCenterX - param.dPitchXY / 2.;
    xdMax = pixelCenterX + param.dPitchXY / 2.;
    ydMin = pixelCenterY - param.dPitchXY / 2.;
    ydMax = pixelCenterY + param.dPitchXY / 2.;
    xsMin = xdMin - param.colD;
    xsMax = xdMax + param.colD;
    ysMin = ydMin - param.colD;
    ysMax = ydMax + param.colD;

    nRay = pow(ceil(sqrt(param.nRaySPECT)), 2);
    detectorShifts = zeros(nRay, 2);
    sourceShifts = zeros(nRay, 2);
    for x = 1:sqrt(nRay)
        for y = 1:sqrt(nRay)
            tmpXd = xdMin + (x - 1) / (sqrt(nRay) - 1) * (xdMax - xdMin);
            tmpYd = ydMin + (y - 1) / (sqrt(nRay) - 1) * (ydMax - ydMin);
            tmpXs = xsMin + (x - 1) / (sqrt(nRay) - 1) * (xsMax - xsMin);
            tmpYs = ysMin + (y - 1) / (sqrt(nRay) - 1) * (ysMax - ysMin);

            detectorShifts((x-1)*sqrt(nRay)+y, :) = [tmpXd tmpYd];
            sourceShifts((x-1)*sqrt(nRay)+y, :) = [tmpXs tmpYs];
        end
    end

    sqShifts = zeros(6, nRay);
    for currentShift = 1:nRay
        xx_d = detectorShifts(currentShift, 1);
        yy_d = detectorShifts(currentShift, 2);
        xx_s = sourceShifts(currentShift, 1);
        yy_s = sourceShifts(currentShift, 2);
            
        if twoDflag
            sqShifts(1:2, currentShift) = [xx_s; yy_s];
            sqShifts(4:5, currentShift) = [xx_d; yy_d];
        else
            sqShifts(1, currentShift) = -sind(panelAngle) * (yy_s-pixelCenterY) - panelNvecX - panelNvecXn * param.colL;
            sqShifts(2, currentShift) = cosd(panelAngle) * (yy_s-pixelCenterY) - panelNvecY - panelNvecYn * param.colL;
            sqShifts(3, currentShift) = xx_s - pixelCenterX;
            sqShifts(4, currentShift) = -sind(panelAngle) * (yy_d-pixelCenterY);
            sqShifts(5, currentShift) = cosd(panelAngle) * (yy_d-pixelCenterY);
            sqShifts(6, currentShift) = xx_d - pixelCenterX;
        end
    end
end

function res = pow(a, b)
    res = a^b;
end